# KamiLimu Data Science Assignment — Python Libraries 1

**Dataset:** Chipotle Orders  
**Libraries used:** `pandas`, `numpy`

### Load Dataset

In [23]:
import pandas as pd
import numpy as np

# Load the Chipotle orders dataset from the remote file
df = pd.read_csv(
    "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv",
    sep='\t'
)

print("Dataset loaded successfully")
print(f"Shape: {df.shape}")

Dataset loaded successfully
Shape: (4622, 5)



## Section 1: Foundational Data Exploration & Manipulation (Qtns 1 to 10)

Q1. Display the last 7 rows of the DataFrame

In [24]:
# tail(n) returns the last n rows of the DF
df.tail(7)

,order_id,quantity,item_name,choice_description,item_price
4615,1832,1,Chicken Soft Tacos,"[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]",$8.75
4616,1832,1,Chips and Guacamole,NaN,$4.45
4617,1833,1,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Black Beans, Sour ...",$11.75
4618,1833,1,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...",$11.75
4619,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$11.25
4620,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Lettu...",$8.75
4621,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$8.75


Q2. Print a summary of the DataFrame

In [25]:
# info() shows index dtype, column dtypes, non-null counts, and memory usage.
df.info()
df.index.dtype

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


dtype('int64')

**Observations from question 2 code:**

- The dataset has 4,622 rows and 5 columns: `order_id`, `quantity`, `item_name`, `choice_description`, and `item_price`.
- `order_id` and `quantity` are stored as **integers** (`int64`), which is ok.
- `item_name`, `choice_description`, and `item_price` are stored as objects (strings).
- `item_price` is a **string** even though it represents monetary value — this is changed in a later question.
- `choice_description` has fewer non-null values than the other columns, indicating some rows have missing descriptions (e.g simple items with no customisation).
- Memory usage is somewhat small ( about 216+ KB), so the dataset is lightweight and easy to work with in memory.

Q3. Determine how many observations have a missing `item_price`

In [26]:
# isnull() returns a boolean; sum() counts the true values
missing_item_price = df['item_price'].isnull().sum()

print(f"Number of observations with missing item_price: {missing_item_price}")

Number of observations with missing item_price: 0


Q4. Print the names of all columns in the DataFrame

In [27]:
# df.columns returns an Index object containing all column names.
column_names = df.columns.tolist()

print("Column names:")
for col in column_names:
    print(f"  - {col}")

Column names:
  - order_id
  - quantity
  - item_name
  - choice_description
  - item_price


 Q5. Identify the most frequently appearing `item_name`

In [28]:
# value_counts() returns item_name frequencies in descending order.
# idxmax() retrieves the index label (item name) of the maximum count.
most_frequent_item = df['item_name'].value_counts().idxmax()
most_frequent_count = df['item_name'].value_counts().max()

print(f"Most frequent item: '{most_frequent_item}' — appears {most_frequent_count} times")

Most frequent item: 'Chicken Bowl' — appears 726 times


 Q6. Determine how many distinct `item_name` values exist

In [29]:
# nunique() counts the number of unique values in the df passed to it
distinct_items = df['item_name'].nunique()

print(f"Number of distinct item names: {distinct_items}")

Number of distinct item names: 50


 Q7. Create a DataFrame containing only rows where `item_name` is 'Chicken Bowl'

In [30]:
# Select rows where the item_name column equals 'Chicken Bowl'.
chicken_bowl_df = df[df['item_name'] == 'Chicken Bowl'].copy()

print(f"Rows with 'Chicken Bowl': {len(chicken_bowl_df)}")
chicken_bowl_df.head()

Rows with 'Chicken Bowl': 726


,order_id,quantity,item_name,choice_description,item_price
4,2,2,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",$16.98
5,3,1,Chicken Bowl,"[Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...",$10.98
13,7,1,Chicken Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Rice,...",$11.25
19,10,1,Chicken Bowl,"[Tomatillo Red Chili Salsa, [Fajita Vegetables...",$8.75
26,13,1,Chicken Bowl,"[Roasted Chili Corn Salsa (Medium), [Pinto Bea...",$8.49


Q8. Convert `item_price` from string to float

In [31]:
#Created a helper function to convert the string prices to float
def parse_price(price_str):
    """
    Convert a price string like '$X.XX' to a float

    Parameters
    ----------
    price_str : str
        A price string that begins with a dollar sign, e.g '$2.39'

    Returns
    -------
    float
        The numeric price value, e.g. 2.39.
    """
    # Remove the leading '$', then turn to float
    return float(price_str.replace('$', '').strip())


# Apply the helper function to all values in the item_price column
# We then overwrite the column in place so the rest of the notebook
# can rely on item_price being a float type
df['item_price'] = df['item_price'].apply(parse_price)

# Confirm if the conversion was successful
print(f"item_price dtype after conversion: {df['item_price'].dtype}")
print("\nSample values:")
print(df['item_price'].head())

item_price dtype after conversion: float64

Sample values:
0     2.39
1     3.39
2     3.39
3     2.39
4    16.98
Name: item_price, dtype: float64


Q9. Calculate the average `item_price` across all orders

In [32]:
# mean() calculates the average of all non-null numeric values in the df
# Now that item_price is a float, this works directly without needing
# to convert further
average_price = df['item_price'].mean()

print(f"Average item price: ${average_price:.2f}")

Average item price: $7.46


Q10. Determine the total number of unique orders

In [33]:
# nunique() on the order_id column gives us the count of distinct order IDs,
# which are the total number of unique orders placed.
unique_orders = df['order_id'].nunique()

print(f"Total number of unique orders: {unique_orders}")

Total number of unique orders: 1834


---
## Section 2: Extra Credit — Total Revenue Calculation (Pandas)

Q1. Calculate Total Revenue

Each row in the dataset represents an item in an order
Revenue formula used = `quantity × item_price` 
Total revenue = sum of revenue across all items

In [34]:
# Frst, create a 'line_revenue' column for each row.
# This multiplies the number of units ordered by the unit price
df['line_revenue'] = df['quantity'] * df['item_price']

# Then, sum all line revenues to get the total revenue across all orders.
total_revenue = df['line_revenue'].sum()

print(f"Total Revenue from all orders: ${total_revenue:,.2f}")

Total Revenue from all orders: $39,237.02


### Assignment ends here. Anything ahead is just for curiosity's sake

In [35]:
# Just to see which items drive the most revenue...
# I think this gives useful insight for the business
revenue_by_item = (
    df.groupby('item_name')['line_revenue']
    .sum()
    .sort_values(ascending=False)
)

print("Top 5 items by revenue:")
print(revenue_by_item.head())
print(f"\nAll items combined: ${revenue_by_item.sum():,.2f}")

Top 5 items by revenue:
item_name
Chicken Bowl           8044.63
Chicken Burrito        6387.06
Steak Burrito          4236.13
Steak Bowl             2479.81
Chips and Guacamole    2475.62
Name: line_revenue, dtype: float64

All items combined: $39,237.02
